### PR TITLE
website: Add Terraform Registry Provider Docs page

### PR DESCRIPTION
**Pull request by justincampbell**
Screen Shot 2019-10-21 at 12 19 03 PM

### CODE REVIEW

File: website/docs/registry/providers/docs.html.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+* Resources
+    * example_cloud
+* Data Sources
+    * example_cloud
+
+Guides are always rendered before Resources, Data Sources, and any subcategories.
+
+### Guides Subcategories
+
+If a provider has many guides, it can be useful to group them into separate top-level folders. These also use the `subcategory` attribute in the YAML frontmatter:
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
If a provider has many guides, you can use subcategories to group them into separate top-level sections. For example, given the following directory structure:
```

File: website/docs/registry/providers/docs.html.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        authenticating-basic.md
+        authenticating-oauth.md
+        setup.md
+    data-sources/
+        cloud.md
+    resources/
+        cloud.md
+```
+
+Given these 3 guides have titles similar to their filename, and `subcategory: "Authentication"` has been added for the first two, the Terraform Registry would display this navigation structure:
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
Assuming that these three guides have titles similar to their filenames, and the first two include `subcategory: "Authentication"` in their frontmatter, the Terraform Registry would display this navigation structure:
```

File: website/docs/registry/providers/docs.html.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    * Initial Setup
+* Authentication
+    * Authenticating with Basic Authentication
+    * Authenticating with OAuth
+* Resources
+    * example_cloud
+* Data Sources
+    * example_cloud
+
+Guides without a subcategory are always rendered before guides with subcategories. Both are always rendered before Resources and Data Sources.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
Guides without a subcategory are always rendered before guides with subcategories. Both are always rendered before resources and data sources.
```

File: website/docs/registry/providers/docs.html.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+## Migrating Legacy Providers Docs
+
+For most provider docs already published to [terraform.io][terraform-io-providers], no changes are required to publish them to the Terraform Registry.
+
+~> The only exception is any providers which organize resources, data sources, or guides into subcategories. See the [Subcategories](#subcategories) section above for more information.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
~> **Important:** The only exceptions are providers which organize resources, data sources, or guides into subcategories. See the [Subcategories](#subcategories) section above for more information.
```

File: website/docs/registry/providers/docs.html.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+For provider docs which are not being published to terraform.io, but wish to be published to the Terraform Registry, take the following steps to migrate to the newer format:
+
+1. Move the `website/docs/` folder to `docs/`
+2. Expand the folder names to be more explicit:
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
2. Expand the folder names to match the Terraform Registry's expected format:
```

File: website/docs/registry/providers/docs.html.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+Guides without a subcategory are always rendered before guides with subcategories. Both are always rendered before Resources and Data Sources.
+
+## Migrating Legacy Providers Docs
+
+For most provider docs already published to [terraform.io][terraform-io-providers], no changes are required to publish them to the Terraform Registry.
+
+~> The only exception is any providers which organize resources, data sources, or guides into subcategories. See the [Subcategories](#subcategories) section above for more information.
+
+For provider docs which are not being published to terraform.io, but wish to be published to the Terraform Registry, take the following steps to migrate to the newer format:
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
If you want to publish docs on the Terraform Registry that are not currently published to terraform.io, take the following steps to migrate to the newer format:
```

File: website/docs/registry/providers/docs.html.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    
+    * List arguments this resource takes.
+    
+    ## Attribute Reference
+    
+    * List attributes that this resource exports.
+
+### YAML Frontmatter
+
+Markdown source files may contain YAML frontmatter, which provides information to the Registry for organization and display of providers.
```
[the remark this answers is not in the document]
Comment (justincampbell) [reply website/docs/registry/providers/docs.html.md]: We only require frontmatter to determine the subcategory, or for guides we use it to set the page_title. Most resources/data sources won't require it.

File: website/docs/registry/providers/docs.html.md [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    
+    * `attribute_name` - (Optional/Required) List arguments this Resource takes.
+    
+    ## Attribute Reference
+    
+    * `attribute_name` - List attributes that this Resource exports.
+
+### YAML Frontmatter
+
+Markdown source files may contain YAML frontmatter, which provides information to the Registry for organization and display of providers.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
Markdown source files may contain YAML frontmatter, which provides organizational information and display hints. Frontmatter can be omitted for resources and data sources that don't require a subcategory. 
```

File: website/docs/registry/providers/docs.html.md [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+Sigil | Text prefix       | Color
+------|-------------------|-------
+`->`  | `**Note**`       | blue
+`~>`  | `**Note**`       | yellow
+`!>`  | `**Warning**`    | red
+
+## Navigation Hierarchy
+
+Provider docs are organized by category: Resources, Data Sources, and Guides. At a minimum, a provider must contain an index (`docs/index.md`) and at least one Resource or Data Source.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
Provider docs are organized by category: resources, data sources, and guides. At a minimum, a provider must contain an index (`docs/index.md`) and at least one resource or data source.
```

File: website/docs/registry/providers/docs.html.md [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+
+### Typical Structure
+
+A provider named `example` with a Resource and Data Source for `cloud` would have these 3 files:
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
A provider named `example` with a resource and data source for `cloud` would have these 3 files:
```

File: website/docs/registry/providers/docs.html.md [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+
+* example Provider
+* Resources
+    * example_cloud
+* Data Sources
+    * example_cloud
+
+### Subcategories
+
+If we wanted to group these Resources by a service or other dimension, we could add a `subcategory` field to the YAML frontmatter of the Resource and Data Source:
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
To group these resources by a service or other dimension, add the optional `subcategory` field to the YAML frontmatter of the resource and data source:
```

File: website/docs/registry/providers/docs.html.md [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+This would change the navigation hierarchy to the following:
+
+* example Provider
+* Cloud
+    * Resources
+        * example_cloud
+    * Data Sources
+        * example_cloud
+
+Resources and Data Sources without a subcategory will be rendered before any subcategories.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
Resources and data sources without a subcategory will be rendered before any subcategories.
```

File: website/docs/registry/providers/docs.html.md [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+
+### Guides
+
+Providers can optionally include 1 or more Guides. These can assist users in using the provider for certain scenarios.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
Providers can optionally include 1 or more guides. These can assist users in using the provider for certain scenarios.
```

File: website/docs/registry/providers/docs.html.md [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+
+### Storage Limits
+
+The maximum number of documents allowed for a single provider version is 1000.
+
+Each document can contain no more than 500KB of data. Documents which exceed this limit will be truncated, and a note will be displayed in the Terraform Registry.
+
+## Format
+
+Provider documentation should be a directory of Markdown documents in the provider repository. Each Markdown document is rendered as a separate page. The directory should include a document for the provider index, a document for each Resource and Data Source, and optional documents for any Guides.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
Provider documentation should be a directory of Markdown documents in the provider repository. Each Markdown document is rendered as a separate page. The directory should include a document for the provider index, a document for each resource and data source, and optional documents for any guides.
```

File: website/docs/registry/providers/docs.html.md [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+| `docs/` | `index.md` | Index page for the provider. |
+| `docs/guides/` | `<guide>.md` | Additional documentation for Guides. |
+| `docs/resources/` | `<resource>.md` | Information for a Resource. Filename should not include a `<PROVIDER NAME>_` prefix. |
+| `docs/data-sources/` | `<data_source>.md` | Information on a provider Data Source. |
+
+-> **Note:** In order to support provider docs which have already been formatted for publishing to [terraform.io][terraform-io-providers], the Terraform Registry also supports docs in a `website/docs/` legacy directory with file extensions of `.html.markdown` or `.html.md`.
+
+### Headers
+
+We strongly suggest that provider docs include the following sections to help users understand how to use the provider. Create additional sections if they would enhance usability of the Resource (for example, “Imports” or “Customizable Timeouts”).
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
We strongly suggest that provider docs include the following sections to help users understand how to use the provider. Create additional sections if they would enhance usability of the resource (for example, “Imports” or “Customizable Timeouts”).
```

File: website/docs/registry/providers/docs.html.md [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+    
+    ## Argument Reference
+    
+    * List any arguments for the provider block.
+
+#### Resources/Data Sources
+
+    # <resource name> Resource/Data Source
+
+    Description of what this Resource does, with links to official app/service documentation.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
    Description of what this resource does, with links to official app/service documentation.
```

File: website/docs/registry/providers/docs.html.md [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+    
+    ## Example Usage
+    
+    ```terraform
+    // Code block with an example of how to use this Resource.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
    // Code block with an example of how to use this resource.
```

File: website/docs/registry/providers/docs.html.md [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+    ```
+    
+    ## Argument Reference
+    
+    * `attribute_name` - (Optional/Required) List arguments this Resource takes.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
    * `attribute_name` - (Optional/Required) List arguments this resource takes.
```

File: website/docs/registry/providers/docs.html.md [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+    
+    ## Attribute Reference
+    
+    * `attribute_name` - List attributes that this Resource exports.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
    * `attribute_name` - List attributes that this resource exports.
```

File: website/docs/registry/providers/docs.html.md [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+subcategory: "Authentication"
+---
+```
+
+#### Supported Attributes
+
+The following frontmatter attributes are supported by the Terraform Registry:
+
+* **page_title** - The title of this document, which will display in the docs navigation. This is only required for documents in the `guides/` folder.
+* **subcategory** - An optional additional layer of grouping that affects the display of the docs navigation; [see Subcategories below](#subcategories) for more details. Resources and Data Sources should be organized into subcategories if the number of Resources would be difficult to quickly scan for a user. Guides should be separated into subcategories if there are multiple Guides which fit into 2 or more distinct groupings.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
* **subcategory** - An optional additional layer of grouping that affects the display of the docs navigation; [see Subcategories below](#subcategories) for more details. Resources and data sources should be organized into subcategories if the number of resources would be difficult to quickly scan for a user. Guides should be separated into subcategories if there are multiple guides which fit into 2 or more distinct groupings.
```

File: website/docs/registry/providers/docs.html.md [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+    index.md
+    Guides/
+        authenticating.md
+    data-sources/
+        cloud.md
+    Resources/
+        cloud.md
+```
+
+The title for Guides is controlled with the `page_title` attribute in the YAML frontmatter:
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
The title for guides is controlled with the `page_title` attribute in the YAML frontmatter:
```

File: website/docs/registry/providers/docs.html.md [hunk 22]
```diff
@@ -0,0 +1,245 @@
...
+* Data Sources
+    * example_cloud
+
+Guides are always rendered before Resources, Data Sources, and any subcategories.
+
+If a `page_title` attribute is not found, the title will default to the filename without the extension.
+
+### Guides Subcategories
+
+If a provider has many Guides, you can use subcategories to group them into separate top-level sections. For example, given the following directory structure:
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
If a provider has many guides, you can use subcategories to group them into separate top-level sections. For example, given the following directory structure:
```

File: website/docs/registry/providers/docs.html.md [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+        authenticating-basic.md
+        authenticating-oauth.md
+        setup.md
+    data-sources/
+        cloud.md
+    Resources/
+        cloud.md
+```
+
+Assuming that these three Guides have titles similar to their filenames, and the first two include `subcategory: "Authentication"` in their frontmatter, the Terraform Registry would display this navigation structure:
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
Assuming that these three guides have titles similar to their filenames, and the first two include `subcategory: "Authentication"` in their frontmatter, the Terraform Registry would display this navigation structure:
```

File: website/docs/registry/providers/docs.html.md [hunk 24]
```diff
@@ -0,0 +1,245 @@
...
+    * Initial Setup
+* Authentication
+    * Authenticating with Basic Authentication
+    * Authenticating with OAuth
+* Resources
+    * example_cloud
+* Data Sources
+    * example_cloud
+
+Guides without a subcategory are always rendered before Guides with subcategories. Both are always rendered before Resources and Data Sources.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
Guides without a subcategory are always rendered before guides with subcategories. Both are always rendered before resources and data sources.
```

File: website/docs/registry/providers/docs.html.md [hunk 25]
```diff
@@ -0,0 +1,245 @@
...
+
+## Migrating Legacy Providers Docs
+
+For most provider docs already published to [terraform.io][terraform-io-providers], no changes are required to publish them to the Terraform Registry.
+
+~> **Important:** The only exceptions are providers which organize Resources, Data Sources, or Guides into subcategories. See the [Subcategories](#subcategories) section above for more information.
```
Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
~> **Important:** The only exceptions are providers which organize resources, data sources, or guides into subcategories. See the [Subcategories](#subcategories) section above for more information.
```

Comment (nfagerlund) [inline website/docs/registry/providers/docs.html.md]:
```suggestion
~> **Important:** The only exceptions are providers which organize resources, data sources, or guides into subcategories. See the [Subcategories](#subcategories) section above for more information.
```

**Pull request merged**